File: engine.py
import json
import os

class AdventureGame:
    def __init__(self, data_file):
        with open(data_file, 'r') as f:
            self.data = json.load(f)
        
        self.current_room = self.data['start_room']
        self.inventory = []
        self.flags = self.data['game_flags']

    def play(self):
        print("--- ADVENTURE START ---")
        while True:
            room = self.data['rooms'][self.current_room]
            print(f"\n{room['description']}")
            
            if room.get('items'):
                print(f"Items here: {', '.join(room['items'])}")

            cmd = input("\n> ").lower().split()
            if not cmd: continue
            
            verb = cmd[0]
            noun = cmd[1] if len(cmd) > 1 else None

            if verb == "go":
                self.move(noun)
            elif verb == "take":
                self.take_item(noun)
            elif verb == "use" or verb == "give":
                self.interact(noun)
            elif verb == "inv":
                print(f"Inventory: {self.inventory}")
            elif verb == "quit":
                break
            else:
                print("I don't know how to do that.")

    def move(self, direction):
        room = self.data['rooms'][self.current_room]
        if direction in room['exits']:
            exit_data = room['exits'][direction]
            
            # Logic Gate Check
            if isinstance(exit_data, dict):
                if self.flags.get(exit_data['required_flag']):
                    self.current_room = exit_data['target']
                else:
                    print(exit_data['fail_msg'])
            else:
                self.current_room = exit_data
        else:
            print("You can't go that way.")

    def take_item(self, item):
        room = self.data['rooms'][self.current_room]
        if item in room.get('items', []):
            self.inventory.append(item)
            room['items'].remove(item)
            print(f"Picked up {item}.")
        else:
            print("That isn't here.")

    def interact(self, noun):
        room = self.data['rooms'][self.current_room]
        # Check interactions defined in world.json
        for action, details in room.get('interactions', {}).items():
            if details['item_required'] in self.inventory:
                self.flags[details['set_flag']] = True
                self.inventory.remove(details['item_required'])
                print(details['success_msg'])
                return
        print("Nothing happened.")

if __name__ == "__main__":
    game = AdventureGame('world.json')
    game.play()

File: engine.js
let gameData = {};
let currentRoom = "";
let inventory = [];
let flags = {};

const display = document.getElementById('display');
const inputField = document.getElementById('cmd-input');

// 1. Load the Game Data
async function initGame() {
    try {
        const response = await fetch('world.json');
        if (!response.ok) throw new Error("Could not find world.json");
        gameData = await response.json();
        
        currentRoom = gameData.start_room;
        flags = gameData.game_flags || {};
        
        renderRoom();
    } catch (err) {
        print("Error: " + err.message + ". Make sure world.json exists in your repo!");
    }
}

// 2. The "Wiring" - Listen for the Enter Key
inputField.addEventListener('keydown', (e) => {
    if (e.key === 'Enter') {
        const val = inputField.value;
        inputField.value = ''; // Clear the input
        processCommand(val);
    }
});

function renderRoom() {
    const room = gameData.rooms[currentRoom];
    print(`<strong>${currentRoom.replace('_', ' ').toUpperCase()}</strong>`);
    print(room.description);
    
    if (room.items && room.items.length > 0) {
        print(`<span style="color: #55FFFF">You see: ${room.items.join(', ')}</span>`);
    }
}

function processCommand(input) {
    const parts = input.toLowerCase().trim().split(' ');
    const verb = parts[0];
    const noun = parts[parts.length - 1]; // Grabs the last word as the object
    const room = gameData.rooms[currentRoom];

    print(`<span class="user-input">> ${input}</span>`);

    if (verb === "go" || verb === "move") {
        movePlayer(noun, room);
    } else if (verb === "take" || verb === "get") {
        takeItem(noun, room);
    } else if (verb === "inv" || verb === "inventory") {
        print("Inventory: " + (inventory.length ? inventory.join(', ') : "Nothing"));
    } else if (verb === "help") {
        print("Commands: GO [direction], TAKE [item], INV, HELP");
    } else {
        print("I don't know how to '" + verb + "'.");
    }
}

function movePlayer(direction, room) {
    if (room.exits && room.exits[direction]) {
        currentRoom = room.exits[direction];
        renderRoom();
    } else {
        print("You can't go that way.");
    }
}

function takeItem(item, room) {
    if (room.items && room.items.includes(item)) {
        inventory.push(item);
        room.items = room.items.filter(i => i !== item);
        print("You took the " + item + ".");
    } else {
        print("That's not here.");
    }
}

function print(text) {
    display.innerHTML += `<p>${text}</p>`;
    display.scrollTop = display.scrollHeight;
}

initGame();
